egui_glow/src/painter.rs

```diff
@@ -3,247 +3,172 @@
+        let header = shader_version.version();
+        glow_debug_print(header);
+        let srgb_support = gl.supported_extensions().contains("EXT_sRGB");
+        let (post_process, srgb_support_define) = match (shader_version, srgb_support) {
+            //WebGL2 support sRGB default
+            (ShaderVersion::Es300, _) | (ShaderVersion::Es100, true) => {
+                //Add sRGB support marker for fragment shader
+                if let Some([width, height]) = pp_fb_extent {
+                    glow_debug_print("WebGL with sRGB enabled so turn on post process");
+                    //install post process to correct sRGB color
+                    (
+                        unsafe { PostProcess::new(gl, is_webgl_1, width, height) }.ok(),
+                        "#define SRGB_SUPPORTED",
+                    )
+                } else {
+                    glow_debug_print("WebGL or OpenGL ES detected but PostProcess disabled because dimension is None");
+                    (None, "")
+                }
             }
+            //WebGL1 without sRGB support disable postprocess and use fallback shader
+            (ShaderVersion::Es100, false) => (None, ""),
+            //OpenGL 2.1 or above always support sRGB so add sRGB support marker
+            _ => (None, "#define SRGB_SUPPORTED"),
+        };
 
-            let program = gl.create_program().unwrap();
-            gl.attach_shader(program, vert);
-            gl.attach_shader(program, frag);
-            gl.link_program(program);
-            if !gl.get_program_link_status(program) {
-                panic!("{}", gl.get_program_info_log(program));
-            }
+        unsafe {
+            let vert = compile_shader(
+                gl,
+                glow::VERTEX_SHADER,
+                &format!(
+                    "{}\n{}\n{}",
+                    header,
+                    shader_version.is_new_shader_interface(),
+                    VERT_SRC
+                ),
+            )
+            .map_err(|problems| {
+                glow_debug_print(format!("failed to compile vertex shader \n {}", problems));
+            })
+            .unwrap();
+            let frag = compile_shader(
+                gl,
+                glow::FRAGMENT_SHADER,
+                &format!(
+                    "{}\n{}\n{}\n{}",
+                    header,
+                    srgb_support_define,
+                    shader_version.is_new_shader_interface(),
+                    FRAG_SRC
+                ),
+            )
+            .map_err(|problems| {
+                glow_debug_print(format!("failed to compile fragment shader \n {}", problems));
+            })
+            .unwrap();
+            let program = link_program(gl, [vert, frag].iter())
+                .map_err(|problems| {
+                    glow_debug_print(format!("failed to link shaders \n {}", problems));
+                })
+                .unwrap();
```

Comment:
Instead of doing

``` rust
.map_err(|problems| {
                    glow_debug_print(format!("failed to link shaders \n {}", problems));
                })
                .unwrap();
```

several times, use `?` and let the caller print the errors (once). i.e. return `Result<Self, String>` from `Painter::new`

Reply:
Ok i fixed it